### PR TITLE
mgr/dashboard: Improve Summary's subscribe methods

### PR DESCRIPTION
**Pull request by tspmelo**
Now subscribe only emits if the value is not undefined,
removing the need to do the validation after each subscribe.

Add a new subscribeOnce method, which replaces getCurrentSummary.
Like the previous one, this will only emit when there is a non undefined value,
and will only emit once.
After it emits, the subscription is closed automatically.

Fixes: https://tracker.ceph.com/issues/36563

Signed-off-by: Tiago Melo <tmelo@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-list.service.ts [hunk 1]
```diff
@@ -57,11 +57,9 @@ export class TaskListService implements OnDestroy {
     this.builders = builders || {};
 
     this.summaryDataSubscription = this.summaryService.subscribe((tasks: any) => {
-      if (tasks) {
-        this.getUpdate().subscribe((resp: any) => {
-          this.updateData(resp, tasks.executing_tasks.filter(this.taskFilter));
-        }, this.onFetchError);
-      }
+      this.getUpdate().subscribe((resp: any) => {
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-list.service.ts:60]: Now that ```if``` condition has been removed we can set the tasks type as object:
```this.summaryService.subscribe((tasks: object)```.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/orchestrator-doc-panel/orchestrator-doc-panel.component.ts [hunk 2]
```diff
@@ -17,17 +17,9 @@ export class OrchestratorDocPanelComponent implements OnInit {
   ) {}
 
   ngOnInit() {
-    const subs = this.summaryService.subscribe((summary: any) => {
-      if (!summary) {
-        return;
-      }
-
+    this.summaryService.subscribeOnce((summary: any) => {
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/orchestrator-doc-panel/orchestrator-doc-panel.component.ts:20]: Here we can define ```summary``` type as ```object```.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/monitoring-list/monitoring-list.component.ts [hunk 3]
```diff
@@ -38,17 +38,9 @@ export class MonitoringListComponent implements OnInit {
       this.isPrometheusConfigured = true;
     });
 
-    const subs = this.summaryService.subscribe((summary: any) => {
-      if (!summary) {
-        return;
-      }
-
+    this.summaryService.subscribeOnce((summary: any) => {
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/monitoring-list/monitoring-list.component.ts:41]: In these cases that we're assuming that ```summary``` payload is an object, we should set the appropriate type.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-list.service.ts (hunk 1 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-list.service.ts:60]: Done.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.spec.ts [hunk 4]
```diff
@@ -66,14 +67,80 @@ describe('SummaryService', () => {
...
+      const subscriber = summaryService.subscribeOnce((response) => {
+        i++;
+        result = response;
+      });
+
+      expect(subscriber).toEqual(jasmine.any(Subscriber));
+      expect(i).toBe(0);
+      expect(result).toEqual(undefined);
+
+      summaryService['summaryDataSource'].next(undefined);
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.spec.ts:87]: How about using a method that is used here instead?

Like
```ts
  const extendSummaryData = (newData: any) => summaryService['summaryDataSource'].next(newData);
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.spec.ts [hunk 5]
```diff
@@ -66,14 +67,80 @@ describe('SummaryService', () => {
...
+      expect(i).toBe(1);
+    });
+
+    it('should call subscribe', () => {
+      let result;
+      let i = 0;
+      const subscriber = summaryService.subscribe((data) => {
+        i++;
+        result = data;
+      });
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.spec.ts:108]: You can set this in a before each in the describe block. Than you could also add a method to expect the state of these three variables, that would reduce the repetitiveness of the test code.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.spec.ts [hunk 6]
```diff
@@ -66,14 +67,80 @@ describe('SummaryService', () => {
...
+
   describe('Should test methods after first refresh', () => {
     beforeEach(() => {
       authStorageService.set('foobar', undefined, undefined);
       summaryService.refresh();
     });
 
-    it('should call getCurrentSummary', () => {
-      expect(summaryService.getCurrentSummary()).toEqual(summary);
+    it('should call subscribeOnce', () => {
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.spec.ts:137]: It looks like an incomplete duplicate of the test above with the same name.

Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.spec.ts:137]: Will remove

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.spec.ts (hunk 5 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.spec.ts]: Done.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.spec.ts (hunk 4 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.spec.ts]: Done.

**Pull request merged**